Issue #937 - Add a server option to limit total allowed throughput

davidBar-On opened the pull request:
* Version of iperf3 (or development branch, such as `master` or
  `3.1-STABLE`) to which this pull request applies:
Master

* Issues fixed (if any): 
#937 

* Brief description of code changes (suitable for use as a commit message):
Add --btotal server option to limit total server's throughput - send and received.
Limit is checked both during a new Client connection - based on the test data rate limit and number of streams, and during the test - based on the actual total average data rate (average is calculated over specified number of throughput report intervals).

bmah888 commented:
Thanks for the pull request!  This is just a basic acknowledgement, I haven't looked at it in detail yet.  I think I understand the basic principle of what you're trying to do and I think it makes sense.

bmah888 commented on src/iperf_api.c:
```diff
@@ -799,6 +836,7 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
         {"udp", no_argument, NULL, 'u'},
         {"bitrate", required_argument, NULL, 'b'},
         {"bandwidth", required_argument, NULL, 'b'},
+	{"btotal", required_argument, NULL, OPT_TOTAL_BITRATE},
```
For this new option flag, what about `server-bitrate-limit` or `server-total-bitrate`?  `btotal` is a bit terse.

bmah888 commented on src/iperf_api.c:
```diff
@@ -2585,6 +2700,12 @@ iperf_reset_test(struct iperf_test *test)
+    test->total_rate_stats_count = 0;
+    test->total_rate_last_interval_index = 0;
+    test->total_rate_traffic_exceeded_limit = 0;
+    for (int i = 0; i < MAX_INTERVAL; i++)
```
Could you move the declaration of `int i` outside the `for()` statement?  It's generating a warning/error on some compilers (notably CentOS 7).

davidBar-On commented:
I have renamed the option to `--server-bitrate-limit`, along with changing related variable names to have "`bitrate_limit`".
Also, moved the `int i` definition out of the `for` loop.

bmah888 commented:
@davidBar-On Thanks, this looks good.  I might be flailing around with the GitHub reviews UI for a bit, but my intent is to merge your PR as it stands now.  We still need a manpage update, but I can take care of that later.  Thanks again!

bmah888 commented:
@davidBar-On We're starting to plan an iperf-3.9 minor release for the early part of August...one of the motivations is to get this feature into a release.  Thanks for your efforts to improve iperf3!
